### ISSUE TITLE
3.1.0 - Simulated (remapped) "Next Track" and "Previous Track" not working

### ISSUE THREAD

**Issue by epjaeger** [MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-#873]
**OS:** 
Windows 10 Pro x64 build 1511

**Issue Descriptions:**
Just tested shortcut keys in the new v3.1.0.  Play/Pause is now working properly, but Next/Prev track still are not working for me whether the app is in focus or not.  I’ve tried all the driver cleaning/reinstalls that I’ve found on reddit and other places on the net without any difference in behavior.  I do not have the Chrome extension installed.

Windows 10 Pro x64 build 1511
Microsoft Natural Ergonomic Keyboard 4000 v1.0
Microsoft Mouse and Keyboard Center 2.7.133.0

I mapped the back/forward keys on the keyboard to next/prev track using the standard software.  The specific settings are:

Browser Back = “Skip to previous media track”
Browser Forward = “Skip to next media track”

**Steps to Reproduce:**
Install MS Mouse and Keyboard Center.  Map the Browser Back key to "Skip to previous media track", and Browser Forward to "Skip to next media track".  Launch GPMDP, play a song, try to use media keys to skip tracks.

**Comment by jostrander** [comment]
Do your keyboard remaps work with say, windows media player or vlc?

**Comment by epjaeger** [comment]
Haven't tested wimp but VLC works fine.

**Comment by jostrander** [comment]
I'm thinking the issue here is the low-level keybindings on windows are registering the events pre-remap since you don't have native media buttons. Have you tried setting the hot keys in the settings using the browser buttons?

**Comment by epjaeger** [comment]
Here's what I've tried so far:
1) Setting them to the default browser back/forward.  This allows me to go forward/back in the GPMDP app both in focus and out of focus (unless I'm focused on another web browser of course).
2) Disable the keys.   This obviously does nothing.
3) Set the keys to next/prev media.  No affect in focus or out of focus.

**In all three cases, I'm unable to bind the keys in the Desktop Settings > Hotkeys menu.**

Now, I did just figure out a workaround.  If I set the back/forward keys to key shortcuts (Ctrl+Shift+N and Ctrl+Shift+M), then they can be set in Desktop Settings > Hotkeys menu.

**Comment by MarshallOfSound** [comment]
I'm on the fence for this one (As in whether it is a bug or not :confused: )

On one hand, you are not getting the expected behavior from the application.

On the other, you are using 3rd party software to trick other applications into thinking you have media keys that you don't :laughing: 

I don't even think there is a way I can fix this, it will come to order of execution of the low level keyboard hooks and apparently mine fires first :cry:

**Comment by epjaeger** [comment]
Third party?  It's the Microsoft keyboard driver.... Wouldn't that be first
party?

**Comment by epjaeger** [comment]
Sorry, that came off snotty and didn't intend for it to.  I figured if
Microsoft maps to the standard "next track" command that works elsewhere it
was working VLC, mpc, and that old outcoldplayer, that was essentially the
same thing.

**Comment by MarshallOfSound** [comment]
It almost is, but also isn't.

From my limited knowledge of how there mapping logic works, they intercept the keypress events at the same level I am listening for them and then cancel that event and fire a new one representing the "simulated" key press or key combination.  The problem is, that event isn't fired at the same level.

And FYI the microsoft keyboard driver is considered third party in this case.  I.e. Not me :+1: 

I can try to look into exactly how this works, but the closest thing I have to test it is my razer keyboard, and those keys are remapped dynamically but the media keys work.  It seems like Microsoft are doing the remapping a strange way.

**Comment by epjaeger** [comment]
Ok got it.  The key combo mapping I made seems to be working fine so I'll
take it for now.  :)

If for some reason you're interested in resolving this or learning more
about the MS behaviors, I do have an old ergo 4000 with a squeaky spacebar
that I could send to you but shipping is damn near as expensive as the
keyboard itself!

**Comment by MarshallOfSound** [comment]
OK so I have looked into this as much as I can and as far as I can tell the Microsoft Mouse and Keyboard Centre application remaps keys at a software level and then sends fake key events to applications.  The problem with this approach is that global hooks method we use for MediaPlayPause does not work at the software level but closer to the driver level (we intercept raw vkCodes).

TLDR: This issue contains a viable (all be it somewhat hacky) workaround to this issue and I don't see an easy way to fix this problem.  It's kind of microsofts fault in the sense that other keyboard remappers I use (Razer and Steelseries just to name 2) work perfectly fine.

**Comment by epjaeger** [comment]
Ok thank you for follow up!   If you have other users in my situation
asking for a fix, just tell them to map the keys to ctrl+shift or ctrl+alt
shortcuts in MS Keyboard Center, and then map those shortcuts in your GPMDP
Hotkeys configuration.  Seems to work just fine for me.

**Comment by MarshallOfSound** [comment]
Will do

Sorry there's not much more I could do
